File: apps/desktop/src/shared/limits.ts
import type { LimitName, Plan, PlanState, UsageMeter } from './cloud'

/**
 * The structured part of a gateway refusal (HTTP 429/413 with `error.limit`), or of the `limit`
 * object `/v1/format` attaches when it skipped the model for fair use. Both apps render these
 * rather than the sentence alone, so the user learns which limit it was and when it comes back.
 * Field names follow internal/entitlements-contract.md in the Project store.
 */
export interface LimitNotice {
  limit: LimitName
  plan: Plan
  planState: PlanState
  used: number
  allowed: number
  /** Epoch ms when the limit next lets a request through; null for a per-request limit. */
  resetsAt: number | null
  /** The web account page that starts an upgrade; null when already Pro or the instance has no site. */
  upgradeUrl: string | null
  /** The gateway's own sentence. */
  message: string
}

const LIMIT_NAMES: ReadonlySet<string> = new Set<LimitName>([
  'wordsPerWeek',
  'sttSecondsPerWeek',
  'dictationsPerDay',
  'maxClipSeconds',
  'sttSecondsPerMonth',
  'fairUseSttSecondsPerMonth',
  'llmTokensPerMonth',
  'requestsPerMinute'
])

export function isLimitName(value: unknown): value is LimitName {
  return typeof value === 'string' && LIMIT_NAMES.has(value)
}

/**
 * Read a limit notice out of an object that carries the contract's fields: a gateway `error`
 * object or a `/v1/format` `limit` object. Null when there is no known `limit` in it.
 */
export function parseLimitNotice(source: unknown, message?: string): LimitNotice | null {
  if (!source || typeof source !== 'object') return null
  const o = source as Record<string, unknown>
  if (!isLimitName(o.limit)) return null
  const plan: Plan = o.plan === 'pro' ? 'pro' : 'free'
  const planState: PlanState =
    o.planState === 'trial' || o.planState === 'pro' || o.planState === 'free' ? o.planState : plan
  return {
    limit: o.limit,
    plan,
    planState,
    used: typeof o.used === 'number' && Number.isFinite(o.used) ? o.used : 0,
    allowed: typeof o.allowed === 'number' && Number.isFinite(o.allowed) ? o.allowed : 0,
    resetsAt: typeof o.resetsAt === 'number' && Number.isFinite(o.resetsAt) ? o.resetsAt : null,
    upgradeUrl: typeof o.upgradeUrl === 'string' && o.upgradeUrl ? o.upgradeUrl : null,
    message: message ?? (typeof o.message === 'string' ? o.message : '')
  }
}

/**
 * The limits that stop a dictation for plan reasons. A rate limit is a plain "try again in a
 * moment" and gets the ordinary error treatment, not the upgrade conversation.
 */
export function isPlanLimit(limit: LimitName): boolean {
  return limit !== 'requestsPerMinute'
}

export function planStateLabel(state: PlanState): string {
  switch (state) {
    case 'trial':
      return 'Pro trial'
    case 'pro':
      return 'Pro'
    default:
      return 'Free'
  }
}

/** The account's plan state, from an instance that reports one or, failing that, from its tier. */
export function planStateOf(
  status: { planState?: PlanState; plan: Plan } | null | undefined,
  user?: { planState?: PlanState; plan: Plan } | null
): PlanState {
  if (status?.planState) return status.planState
  if (user?.planState) return user.planState
  const plan = status?.plan ?? user?.plan ?? 'free'
  return plan === 'pro' ? 'pro' : 'free'
}

/** Whole days left on the trial, never negative: `ceil((trialEndsAt - now) / 24 h)`. */
export function trialDaysLeft(trialEndsAt: number | null | undefined, now = Date.now()): number {
  if (typeof trialEndsAt !== 'number') return 0
  return Math.max(0, Math.ceil((trialEndsAt - now) / 86_400_000))
}

// ---- wording --------------------------------------------------------------------------------

const MINUTE = 60_000
const HOUR = 3_600_000
const DAY = 86_400_000
const WEEKDAYS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

function clockTime(d: Date): string {
  const h = d.getHours()
  const suffix = h < 12 ? 'am' : 'pm'
  const hour = h % 12 === 0 ? 12 : h % 12
  return `${hour}:${String(d.getMinutes()).padStart(2, '0')} ${suffix}`
}

function sameLocalDay(a: Date, b: Date): boolean {
  return (
    a.getFullYear() === b.getFullYear() &&
    a.getMonth() === b.getMonth() &&
    a.getDate() === b.getDate()
  )
}

/**
 * When a limit comes back, the way a person would say it: "in 40 min", "at 3:00 pm",
 * "tomorrow at 1:00 am", "Tue 16 Sep". Local time; the gateway's instants are UTC-aligned.
 */
export function formatResetTime(resetsAt: number, now = Date.now()): string {
  const diff = resetsAt - now
  if (diff < MINUTE) return 'in a moment'
  if (diff < HOUR) return `in ${Math.ceil(diff / MINUTE)} min`
  const then = new Date(resetsAt)
  const today = new Date(now)
  if (sameLocalDay(then, today)) return `at ${clockTime(then)}`
  const tomorrow = new Date(now + DAY)
  if (sameLocalDay(then, tomorrow)) return `tomorrow at ${clockTime(then)}`
  const date = `${then.getDate()} ${MONTHS[then.getMonth()]}`
  return diff < 6 * DAY ? `${WEEKDAYS[then.getDay()]} ${date}` : `on ${date}`
}

function formatCount(n: number): string {
  return new Intl.NumberFormat('en-US', { maximumFractionDigits: 0 }).format(n)
}

/** Token counts read in thousands and millions: "12k", "500k", "2.1M", "25M". */
export function compactCount(n: number): string {
  if (n >= 1_000_000) {
    const m = n / 1_000_000
    return `${m >= 10 || Number.isInteger(m) ? Math.round(m) : m.toFixed(1)}M`
  }
  if (n >= 10_000) return `${Math.round(n / 1000)}k`
  return formatCount(n)
}

/** Seconds of audio as a figure in the unit the allowance is naturally read in: minutes, or hours from 3 h up. */
function audioFigure(seconds: number, allowedSeconds: number): { value: string; unit: string } {
  const hours = allowedSeconds >= 3 * 3600 && allowedSeconds % 3600 === 0
  if (hours) {
    const h = seconds / 3600
    const value = h >= 10 || Number.isInteger(h) ? Math.round(h).toString() : h.toFixed(1)
    return { value, unit: 'h' }
  }
  const m = seconds / 60
  return { value: m > 0 && m < 1 ? '<1' : Math.round(m).toString(), unit: 'min' }
}

/** "7 min", "120 min", "30 h"; `allowedSeconds` decides the unit when a used figure is shown against it. */
export function formatAudioSeconds(seconds: number, allowedSeconds = seconds): string {
  const { value, unit } = audioFigure(seconds, allowedSeconds)
  return `${value} ${unit}`
}

/** "312 of 500 words", "12 of 120 min", "2.5 of 30 h", "3 of 12". */
export function meterValue(meter: Pick<UsageMeter, 'limit' | 'used' | 'allowed'>): string {
  switch (meter.limit) {
    case 'wordsPerWeek':
      return `${formatCount(meter.used)} of ${formatCount(meter.allowed)} words`
    case 'sttSecondsPerWeek':
    case 'sttSecondsPerMonth':
    case 'fairUseSttSecondsPerMonth':
      return `${audioFigure(meter.used, meter.allowed).value} of ${formatAudioSeconds(meter.allowed)}`
    case 'llmTokensPerMonth':
      return `${compactCount(meter.used)} of ${compactCount(meter.allowed)} tokens`
    case 'maxClipSeconds':
      return `up to ${formatAudioSeconds(meter.allowed)} a clip`
    default:
      return `${formatCount(meter.used)} of ${formatCount(meter.allowed)}`
  }
}

/** The Account page's label for a meter. */
export function meterLabel(limit: LimitName): string {
  switch (limit) {
    case 'wordsPerWeek':
      return 'Words this week'
    case 'sttSecondsPerWeek':
      return 'Speech this week'
    case 'dictationsPerDay':
      return 'Dictations today'
    case 'sttSecondsPerMonth':
      return 'Transcription this month'
    case 'fairUseSttSecondsPerMonth':
      return 'Fair use this month'
    case 'llmTokensPerMonth':
      return 'Formatting this month'
    case 'maxClipSeconds':
      return 'Recording length'
    default:
      return 'Requests this minute'
  }
}

/** Meters worth a row on the Account page: the ones that fill up over time. */
export function usageMeters(meters: UsageMeter[] | undefined): UsageMeter[] {
  return (meters ?? []).filter(
    (m) => m.limit !== 'maxClipSeconds' && m.limit !== 'requestsPerMinute'
  )
}

export interface LimitCopy {
  /** One line: what ran out. */
  title: string
  /** One line: the allowance, and when it comes back. */
  detail: string
  /** Upgrading lifts this limit (never for a cap Pro already has). */
  upgradeHelps: boolean
}

function tierPhrase(notice: LimitNotice): string {
  return notice.plan === 'pro' ? 'on Pro' : 'on the free plan'
}

/**
 * The calm version of a limit refusal: which limit it was, what the allowance is, and when it
 * resets. `stage` says what the limit stopped; a formatting limit never loses text (the rule-based
 * cleanup is inserted instead), so its copy leads with that and gives the reason second.
 */
export function describeLimit(
  notice: LimitNotice,
  now = Date.now(),
  stage: 'speech' | 'formatting' = 'speech'
): LimitCopy {
  const reset = notice.resetsAt !== null ? formatResetTime(notice.resetsAt, now) : null
  const resets = reset ? ` · resets ${reset}` : ''
  const stop = describeStop(notice)
  if (stage === 'formatting') {
    return {
      title: 'Inserted without formatting',
      detail: `${stop.title}${resets}`,
      upgradeHelps: stop.upgradeHelps
    }
  }
  return { ...stop, detail: `${stop.detail}${resets}` }
}

/** What ran out and the allowance behind it, without the reset (added by the caller). */
function describeStop(notice: LimitNotice): LimitCopy {
  const tier = tierPhrase(notice)
  const pro = notice.plan === 'pro'
  switch (notice.limit) {
    case 'wordsPerWeek':
      return {
        title: "This week's free words are used up",
        detail: `${formatCount(notice.allowed)} words a week ${tier}`,
        upgradeHelps: true
      }
    case 'sttSecondsPerWeek':
      return {
        title: "This week's free minutes are used up",
        detail: `${formatAudioSeconds(notice.allowed)} of speech a week ${tier}`,
        upgradeHelps: true
      }
    case 'dictationsPerDay':
      return {
        title: "Today's free dictations are used up",
        detail: `${formatCount(notice.allowed)} dictations a day ${tier}`,
        upgradeHelps: true
      }
    case 'maxClipSeconds':
      return {
        title: pro ? 'That recording is too long' : 'That recording is too long for the free plan',
        detail: pro
          ? `Clips can be up to ${formatAudioSeconds(notice.allowed)}`
          : `Clips up to ${formatAudioSeconds(notice.allowed)} ${tier}; up to 10 min on Pro`,
        upgradeHelps: !pro
      }
    case 'sttSecondsPerMonth':
      return {
        title: pro
          ? "This month's fair-use cap is reached"
          : "This month's free transcription is used up",
        detail: `${formatAudioSeconds(notice.allowed)} a month ${tier}`,
        upgradeHelps: !pro
      }
    case 'fairUseSttSecondsPerMonth':
      return {
        title: 'Fair use reached for this month',
        detail: `Past ${formatAudioSeconds(notice.allowed)} of transcription a month the text is tidied by rules only`,
        upgradeHelps: false
      }
    case 'llmTokensPerMonth':
      return {
        title: "This month's formatting allowance is used up",
        detail: `${compactCount(notice.allowed)} tokens a month ${tier}`,
        upgradeHelps: !pro
      }
    default:
      return {
        title: 'Too many requests at once',
        detail: `Up to ${formatCount(notice.allowed)} requests a minute ${tier}`,
        upgradeHelps: !pro
      }
  }
}

File: apps/android/app/src/main/java/app/murmur/android/inference/Limits.kt
package app.murmur.android.inference

import app.murmur.android.cloud.InferenceStatusDto
import app.murmur.android.cloud.UsageMeterDto
import app.murmur.android.cloud.UserDto
import org.json.JSONObject
import java.util.Calendar
import java.util.Locale
import kotlin.math.ceil
import kotlin.math.roundToInt

/**
 * The structured part of a gateway refusal (HTTP 429/413 with `error.limit`), or of the `limit`
 * object `/v1/format` attaches when it skipped the model for fair use. The pill renders these
 * rather than the sentence alone, so the user learns which limit it was and when it comes back.
 * Port of apps/desktop/src/shared/limits.ts; field names follow the entitlements contract.
 */
data class LimitNotice(
    val limit: String,
    val plan: String,
    val planState: String,
    val used: Double,
    val allowed: Double,
    /** Epoch ms when the limit next lets a request through; null for a per-request limit. */
    val resetsAt: Long?,
    /** The web account page that starts an upgrade; null when already Pro or the instance has no site. */
    val upgradeUrl: String?,
    /** The gateway's own sentence. */
    val message: String
) {
    /** A limit that stops a dictation for plan reasons; a rate limit is a plain "try again in a moment". */
    val isPlanLimit: Boolean get() = Limits.isPlanLimit(limit)

    companion object {
        /**
         * Read a notice out of an object carrying the contract's fields (a gateway `error` object or
         * a `/v1/format` `limit` object). Null when there is no known `limit` in it.
         */
        fun fromJson(source: JSONObject?, message: String? = null): LimitNotice? {
            if (source == null) return null
            val limit = source.optString("limit", "")
            if (limit !in Limits.NAMES) return null
            val plan = if (source.optString("plan") == "pro") "pro" else "free"
            val planState = source.optString("planState").takeIf { it in Limits.PLAN_STATES } ?: plan
            return LimitNotice(
                limit = limit,
                plan = plan,
                planState = planState,
                used = source.optDouble("used", 0.0).takeIf { it.isFinite() } ?: 0.0,
                allowed = source.optDouble("allowed", 0.0).takeIf { it.isFinite() } ?: 0.0,
                resetsAt = if (source.isNull("resetsAt") || !source.has("resetsAt")) null else source.optDouble("resetsAt").takeIf { it.isFinite() }?.toLong(),
                upgradeUrl = source.optString("upgradeUrl", "").takeIf { !source.isNull("upgradeUrl") && it.isNotEmpty() },
                message = message ?: source.optString("message", "")
            )
        }
    }
}

/** How a limit is worded: what ran out, the allowance and its reset, whether upgrading helps. */
data class LimitCopy(val title: String, val detail: String, val upgradeHelps: Boolean)

/** What the limit stopped: a formatting limit never loses text (rules are applied instead). */
enum class LimitStage { SPEECH, FORMATTING }

object Limits {
    val NAMES = setOf(
        "wordsPerWeek", "sttSecondsPerWeek", "dictationsPerDay", "maxClipSeconds",
        "sttSecondsPerMonth", "fairUseSttSecondsPerMonth", "llmTokensPerMonth", "requestsPerMinute"
    )
    val PLAN_STATES = setOf("trial", "free", "pro")

    private const val MINUTE = 60_000L
    private const val HOUR = 3_600_000L
    private const val DAY = 86_400_000L
    private val WEEKDAYS = listOf("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")
    private val MONTHS = listOf("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")

    fun isPlanLimit(limit: String): Boolean = limit != "requestsPerMinute"

    fun planStateLabel(state: String): String = when (state) {
        "trial" -> "Pro trial"
        "pro" -> "Pro"
        else -> "Free"
    }

    /** "Pro trial", "Free plan", "Pro plan": the plan as a title. */
    fun planTitle(state: String): String = if (state == "trial") "Pro trial" else "${planStateLabel(state)} plan"

    /** The account's plan state, from an instance that reports one or, failing that, from its tier. */
    fun planStateOf(status: InferenceStatusDto?, user: UserDto?): String {
        status?.planState?.takeIf { it in PLAN_STATES }?.let { return it }
        user?.planState?.takeIf { it in PLAN_STATES }?.let { return it }
        val plan = status?.plan ?: user?.plan ?: "free"
        return if (plan == "pro") "pro" else "free"
    }

    /** Whole days left on the trial, never negative: `ceil((trialEndsAt - now) / 24 h)`. */
    fun trialDaysLeft(trialEndsAt: Double?, now: Long = System.currentTimeMillis()): Int {
        if (trialEndsAt == null || !trialEndsAt.isFinite()) return 0
        return ceil((trialEndsAt - now) / DAY).toInt().coerceAtLeast(0)
    }

    private fun clockTime(cal: Calendar): String {
        val h = cal.get(Calendar.HOUR_OF_DAY)
        val hour = if (h % 12 == 0) 12 else h % 12
        return "%d:%02d %s".format(Locale.US, hour, cal.get(Calendar.MINUTE), if (h < 12) "am" else "pm")
    }

    private fun sameLocalDay(a: Calendar, b: Calendar): Boolean =
        a.get(Calendar.YEAR) == b.get(Calendar.YEAR) && a.get(Calendar.DAY_OF_YEAR) == b.get(Calendar.DAY_OF_YEAR)

    /**
     * When a limit comes back, the way a person would say it: "in 40 min", "at 3:00 pm",
     * "tomorrow at 1:00 am", "Tue 16 Sep", "on 25 Sep". Local time.
     */
    fun formatResetTime(resetsAt: Long, now: Long = System.currentTimeMillis()): String {
        val diff = resetsAt - now
        if (diff < MINUTE) return "in a moment"
        if (diff < HOUR) return "in ${ceil(diff / MINUTE.toDouble()).toInt()} min"
        val then = Calendar.getInstance().apply { timeInMillis = resetsAt }
        val today = Calendar.getInstance().apply { timeInMillis = now }
        if (sameLocalDay(then, today)) return "at ${clockTime(then)}"
        val tomorrow = Calendar.getInstance().apply { timeInMillis = now + DAY }
        if (sameLocalDay(then, tomorrow)) return "tomorrow at ${clockTime(then)}"
        val date = "${then.get(Calendar.DAY_OF_MONTH)} ${MONTHS[then.get(Calendar.MONTH)]}"
        return if (diff < 6 * DAY) "${WEEKDAYS[then.get(Calendar.DAY_OF_WEEK) - 1]} $date" else "on $date"
    }

    fun formatCount(n: Double): String = "%,d".format(Locale.US, n.roundToInt())

    /** Token counts read in thousands and millions: "12k", "500k", "2.1M", "25M". */
    fun compactCount(n: Double): String = when {
        n >= 1_000_000 -> {
            val m = n / 1_000_000
            if (m >= 10 || m == Math.floor(m)) "${m.roundToInt()}M" else "%.1fM".format(Locale.US, m)
        }
        n >= 10_000 -> "${(n / 1000).roundToInt()}k"
        else -> formatCount(n)
    }

    private fun audioFigure(seconds: Double, allowedSeconds: Double): Pair<String, String> {
        val hours = allowedSeconds >= 3 * 3600 && allowedSeconds % 3600.0 == 0.0
        if (hours) {
            val h = seconds / 3600
            val value = if (h >= 10 || h == Math.floor(h)) h.roundToInt().toString() else "%.1f".format(Locale.US, h)
            return value to "h"
        }
        val m = seconds / 60
        return (if (m > 0 && m < 1) "<1" else m.roundToInt().toString()) to "min"
    }

    /** "7 min", "120 min", "30 h"; [allowedSeconds] decides the unit when a used figure is shown against it. */
    fun formatAudioSeconds(seconds: Double, allowedSeconds: Double = seconds): String {
        val (value, unit) = audioFigure(seconds, allowedSeconds)
        return "$value $unit"
    }

    /** "312 of 500 words", "12 of 120 min", "2.5 of 30 h", "3 of 12". */
    fun meterValue(limit: String, used: Double, allowed: Double): String = when (limit) {
        "wordsPerWeek" -> "${formatCount(used)} of ${formatCount(allowed)} words"
        "sttSecondsPerWeek", "sttSecondsPerMonth", "fairUseSttSecondsPerMonth" ->
            "${audioFigure(used, allowed).first} of ${formatAudioSeconds(allowed)}"
        "llmTokensPerMonth" -> "${compactCount(used)} of ${compactCount(allowed)} tokens"
        "maxClipSeconds" -> "up to ${formatAudioSeconds(allowed)} a clip"
        else -> "${formatCount(used)} of ${formatCount(allowed)}"
    }

    fun meterValue(meter: UsageMeterDto): String = meterValue(meter.limit, meter.used, meter.allowed)

    /** The Account screen's label for a meter. */
    fun meterLabel(limit: String): String = when (limit) {
        "wordsPerWeek" -> "Words this week"
        "sttSecondsPerWeek" -> "Speech this week"
        "dictationsPerDay" -> "Dictations today"
        "sttSecondsPerMonth" -> "Transcription this month"
        "fairUseSttSecondsPerMonth" -> "Fair use this month"
        "llmTokensPerMonth" -> "Formatting this month"
        "maxClipSeconds" -> "Recording length"
        else -> "Requests this minute"
    }

    /** Meters worth a row on the Account screen: the ones that fill up over time. */
    fun usageMeters(meters: List<UsageMeterDto>): List<UsageMeterDto> =
        meters.filter { it.limit in NAMES && it.limit != "maxClipSeconds" && it.limit != "requestsPerMinute" }

    /**
     * The calm version of a refusal: which limit it was, the allowance, and when it resets. A
     * formatting limit never loses text (the rule-based cleanup is inserted instead), and the
     * wording says so.
     */
    fun describe(notice: LimitNotice, now: Long = System.currentTimeMillis(), stage: LimitStage = LimitStage.SPEECH): LimitCopy {
        val reset = notice.resetsAt?.let { formatResetTime(it, now) }
        val resets = if (reset != null) " · resets $reset" else ""
        val pro = notice.plan == "pro"
        val tier = if (pro) "on Pro" else "on the free plan"
        val unformatted = stage == LimitStage.FORMATTING
        return when (notice.limit) {
            "wordsPerWeek" -> LimitCopy(
                "This week's free words are used up",
                "${formatCount(notice.allowed)} words a week $tier$resets",
                true
            )
            "sttSecondsPerWeek" -> LimitCopy(
                "This week's free minutes are used up",
                "${formatAudioSeconds(notice.allowed)} of speech a week $tier$resets",
                true
            )
            "dictationsPerDay" -> LimitCopy(
                if (unformatted) "Inserted without formatting: today's free dictations are used up" else "Today's free dictations are used up",
                "${formatCount(notice.allowed)} dictations a day $tier$resets",
                true
            )
            "maxClipSeconds" -> LimitCopy(
                if (pro) "That recording is too long" else "That recording is too long for the free plan",
                if (pro) "Clips can be up to ${formatAudioSeconds(notice.allowed)}"
                else "Clips up to ${formatAudioSeconds(notice.allowed)} $tier; up to 10 min on Pro",
                !pro
            )
            "sttSecondsPerMonth" -> LimitCopy(
                if (pro) "This month's transcription has reached its fair-use cap" else "This month's free transcription is used up",
                "${formatAudioSeconds(notice.allowed)} a month $tier$resets",
                !pro
            )
            "fairUseSttSecondsPerMonth" -> LimitCopy(
                "Inserted without formatting: fair use reached for this month",
                "Past ${formatAudioSeconds(notice.allowed)} of transcription a month the text is tidied by rules only$resets",
                false
            )
            "llmTokensPerMonth" -> LimitCopy(
                if (unformatted) "Inserted without formatting: this month's formatting allowance is used up" else "This month's formatting allowance is used up",
                "${compactCount(notice.allowed)} tokens a month $tier$resets",
                !pro
            )
            else -> LimitCopy(
                "Too many requests at once",
                "Up to ${formatCount(notice.allowed)} requests a minute $tier$resets",
                !pro
            )
        }
    }
}
